File: core/src/de/project/ice/ecs/components/AnimationComponent.java
package de.project.ice.ecs.components;

import com.badlogic.gdx.utils.IntMap;
import de.project.ice.utils.Assets;
import org.jetbrains.annotations.NotNull;

public class AnimationComponent implements IceComponent<AnimationComponent>
{
    public int animation = 0;
    public float time = 0.0f;

    @NotNull
    public IntMap<Assets.AnimationHolder> animations = new IntMap<Assets.AnimationHolder>();

    @Override
    public void reset()
    {
        animations = new IntMap<Assets.AnimationHolder>();
        animation = 0;
        time = 0.0f;
    }

    public int getAnimation()
    {
        return animation;
    }

    public void setAnimation(int animation)
    {
        this.animation = animation;
        time = 0.0f;
    }

    @Override
    public void copyTo(AnimationComponent copy)
    {
        copy.animation = animation;
        for (IntMap.Entry<Assets.AnimationHolder> animationHolder : animations)
        {
            copy.animations.put(animationHolder.key,
                    Assets.createAnimation(
                            animationHolder.value.name,
                            animationHolder.value.data.getFrameDuration(),
                            animationHolder.value.data.getPlayMode()
                    )
            );
        }
        copy.time = time;
    }
}

File: core/src/de/project/ice/ecs/components/AnimationComponent.kt
package de.project.ice.ecs.components

import com.badlogic.gdx.utils.IntMap
import de.project.ice.annotations.Property
import de.project.ice.utils.Assets

class AnimationComponent : CopyableIceComponent {
    @Property("The active animation")
    var animation = 0
        set(animation) {
            field = animation
            time = 0.0f
        }
    @Property("The current animation time", true)
    var time = 0.0f

    @Property("The list of animations")
    val animations = IntMap<Assets.Holder.Animation>()

    override fun reset() {
        animations.clear()
        animation = 0
        time = 0.0f
    }

    override fun copyTo(copy: CopyableIceComponent) {
        if (copy is AnimationComponent) {
            copy.animation = animation
            for (animationHolder in animations) {
                val data = animationHolder.value.data
                if (data != null) {
                    copy.animations.put(animationHolder.key,
                            Assets.createAnimation(
                                    animationHolder.value.name,
                                    data.frameDuration,
                                    data.playMode))
                }
            }
            copy.time = time
        }
    }
}